pallets/pallet-deposits/src/lib.rs

```diff
@@ -0,0 +1,270 @@
+		DepositTakeAnnouncement<<T as frame_system::Config>::BlockNumber, BalanceOf<T>>,
+	>;
+
+	/// Event definitions
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A deposit was paid (deposit id + account id + new total amount)
+		DepositPaid(DepositIdOf<T>, AccountIdOf<T>, BalanceOf<T>),
+		/// A deposit is about to be withdrawn 
+		DepositTakeAnnouncement(
+			DepositIdOf<T>, 
+			AccountIdOf<T>, 
+			<T as frame_system::Config>::BlockNumber,
+			BalanceOf<T>,
+		),
+		/// A deposit was taken (deposit id + account id + new total amount)
+		DepositTaken(DepositIdOf<T>, AccountIdOf<T>, BalanceOf<T>),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		InsufficientFunds,
+		NoticePeriodTooShort,
+		NotAnnounced,
+		StillInNoticePeriod,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Pay a deposit
+		/// # Arguments
+		/// * `origin`: The account that is paying the deposit
+		/// * `deposit_id` - The id of the deposit
+		/// * `amount` - The amount to pay
+		#[pallet::weight(<T as Config>::WeightInfo::pay())]
+		pub fn pay(origin: OriginFor<T>, deposit_id: DepositIdOf<T>, amount: BalanceOf<T>) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+
+			// check if sender has enough funds
+			ensure!(
+				<T::Currency as ReservableCurrency<AccountIdOf<T>>>::can_reserve(&sender, amount,),
+				Error::<T>::InsufficientFunds
+			);
+
+			// reserve the deposit
+			<T::Currency as ReservableCurrency<AccountIdOf<T>>>::reserve(&sender, amount)?;
+
+			// update the account balance for this deposit pot
+			DepositsByAccount::<T>::mutate(&deposit_id, &sender, |balance| {
+				*balance = Some(balance.unwrap_or_default().saturating_add(amount))
+			});
+
+			// update the total amount for this deposit pot
+			let new_total = DepositsTotal::<T>::mutate(&deposit_id, |total| {
+				*total = Some(total.unwrap_or_default().saturating_add(amount));
+				total.unwrap() // safe because we make it Some() one line above
+			});
+
+			// call the change handler with the new total
+			<T::ChangeHandler>::on_deposit_paid(&deposit_id, &sender, new_total);
+
+			// emit event
+			Self::deposit_event(Event::DepositPaid(deposit_id, sender, new_total));
+
+			Ok(())
+		}
+
+		#[pallet::weight(<T as Config>::WeightInfo::announce_take())]
+		pub fn announce_take(
+			origin: OriginFor<T>,
+			deposit_id: DepositIdOf<T>,
+			amount: BalanceOf<T>,
+			block: <T as frame_system::Config>::BlockNumber,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			
+			// check minimal notice period
+			let current_block = frame_system::Pallet::<T>::block_number();
+			if block.saturating_sub(current_block) < T::MinimalNoticePeriod::get() {
+				return Err(Error::<T>::NoticePeriodTooShort.into());
+			}
+
+			// create the announcement
+			let announcment = DepositTakeAnnouncement {
+				block,
+				amount,
+			};
+			Announcments::<T>::insert(&deposit_id, &sender,  announcment);
```

Comment:
This is also consuming storage space, so there should probably be another deposit which is released when the original deposit is actually taken.